dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/action/validation/BaseProgramRuleActionValidator.java

```diff
@@ -193,4 +189,93 @@ private ProgramRuleActionValidationResult handleTrackedEntityAttribute(
 
     return ProgramRuleActionValidationResult.builder().valid(true).build();
   }
+
+  public List<ProgramStage> getRequiredProgramStages(
+      ProgramRuleActionValidationContext validationContext, Program program) {
+    if (validationContext == null || program == null) {
+      return Collections.emptyList();
+    }
+
+    List<ProgramStage> availableStages = validationContext.getProgramStages();
```

Comment:
Can `availableStages` be marked as @Nonnull in `ProgramRuleActionValidationContext` ?